source/agora/consensus/protocol/Nominator.d

```diff
@@ -1031,24 +1033,86 @@ extern(D):
     public override Value combineCandidates (uint64_t slot_idx,
         ref const(set!Value) candidates)
     {
+        import agora.common.Amount;
+
+        /// Used for holding consensus candidate values. It contains precomputed
+        /// fields to speed up sorting.
+        static struct CandidateHolder
+        {
+            /// Consensus data
+            public ConsensusData consensus_data;
+            /// Hash of the consensus data
+            public Hash hash;
+            /// The total amount of fees of the transactions in the consensus data
+            public Amount total_adjusted_fee;
+
+            /// Comparison function, which sorts by
+            /// 1. length of missing validators (smallest first), or if it ties
+            /// 2. total adjusted fee (smallest last), or if it ties
+            /// 3. hash of the candidate (smallest first)
+            public int opCmp (in CandidateHolder other) const @safe scope pure nothrow @nogc
+            {
+                if (this.consensus_data.missing_validators.length <
+                    other.consensus_data.missing_validators.length)
+                        return -1;
+                    else if (this.consensus_data.missing_validators.length >
+                             other.consensus_data.missing_validators.length)
+                        return 1;
+
+                    if (this.total_adjusted_fee > other.total_adjusted_fee)
+                        return -1;
+                    else if (this.total_adjusted_fee < other.total_adjusted_fee)
+                        return 1;
+
+                    if (this.hash < other.hash)
+                        return -1;
+                    else if (this.hash > other.hash)
+                        return 1;
+
+                    return 0;
+            }
+        }
+
         try
         {
-            ConsensusData[] values;
+            CandidateHolder[] candidate_holders;
             foreach (ref const(Value) candidate; candidates)
             {
                 auto data = deserializeFull!ConsensusData(candidate[]);
+                log.trace("Consensus data: {}", data.prettify);
 
+                // Only allowed in unittests, as validating the consensus data
+                // for all the candidates might take long.
+                version (unittest)
```

Comment:
Not so convinced about this. Did you observe any problem ?

Reply:
https://github.com/bosagora/agora/issues/1988#issuecomment-823977927

I don't see the reason validating every single candididates over and over again during the combine phase. In an other comment below I mentioned that from the SCP code it looks we have to validate the ConsensusData first to become a candidate.

Reply:
Faire enough